streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -2494,6 +2495,42 @@ class ZStream[-R, +E, +A](val channel: ZChannel[R, Any, Any, Any, E, Chunk[A], A
         }
     }
 
+  /**
+   * Converts an unordered stream where each element is zipped with its index
+   * into an ordered stream.
+   */
+  final def ordered[A1](implicit ev: A <:< (A1, Long), trace: ZTraceElement): ZStream[R, E, A1] =
+    ZStream.suspend {
+      val ord           = Ordering.by[(A1, Long), Long](_._2).reverse
+      val priorityQueue = PriorityQueue[(A1, Long)]()(ord)
+      var n             = 0L
```

Comment:
Should this be a parameter by default `0L`?

Reply:
I do think we should change this to be configurable. It would be better to avoid the need for it, but I can't think of a robust way to do that.

Reply:
Maybe `orderedFrom(0L)` versus `ordered` (which `n = 0L`).

Reply:
One use case that comes to mind are records coming of Kafka. The orderedFrom variant could even take a value of any ordered type? Argh no, there is nothing like `Enum` 😒

Reply:
The only problem is if you do something like `stream.zipWithIndex.drop(1).ordered`, it automatically will deadlock forever, which is maybe surprising. It's possibly a sign we need a scoped operator that can feed through the index and order automatically. Maybe such an operator would take a `Pipeline` and it would thread the index through automatically, pushing all concurrency to pipelines.

Anyway, we can figure that out later, I'd say we should merge for now and revise later.

Reply:
I think there is an issue here around back pressure. Consider the following test:

```scala
test("back pressure") {
  val stream = ZStream.range(1, 100).rechunk(1).tap(n => ZIO.debug(n)).mapZIOPar(8) { n =>
    if (n == 1) ZIO.never *> ZIO.succeed(1)
    else ZIO.succeed(n)  
  }
  for {
    _ <- stream.take(1).runCollect
  } yield assertCompletes
}
```

If the first item is very slow, the current implementation will back pressure until it completes. In this implementation on the other hand, the `ordered` implementation will keep pulling until it gets a value that it can emit, potentially building up a very large buffer. Furthermore, we can't just limit the buffer size because we could get into a situation where the buffer is filled with values other than the one we need and we deadlock.

It seems like we potentially need a more powerful data structure that can handle back pressure or we need to handle that at a higher level. Perhaps we use a `Semaphore` and `acquire` a permit every time we we start doing a task upstream and release it when we take it downstream.

Reply:
Great spot! Need to be super careful about overhead here since this combinator will be used so often that performance is of utmost importance.